First draft for static validation (#260)
Kattis/problem-package-format: spec/2023-07-draft.md

`scoring`               | Map                                 | [See Verdict/Score Aggregation](#verdict/score-aggregation) | Not inherited  | Description of how the results of the group test cases and subgroups should be aggregated. This key is only permitted for the `secret` group and its subgroups.
`input_validator_args`  | String or map of strings to strings | empty string                            | Inherited      | Arguments passed to each input validator for this test data group. If a string then those are the flags that will be passed to each input validator for this test data group. If a map then each key is the name of the input validator and the value is the flags to pass to that input validator for this test data group. Validators not present in the map are run without flags.
`output_validator_args` | String                              | empty string                            | Inherited      | Arguments passed to the output validator for this test data group.

> | static_validation      | Map or boolean                      | false                   | Not applicable | Configuration of static validation test data node. See [Static Validation](#static-validation)

Reviewer: ```suggestion
`static_validation`    | Map or boolean                      | false                   | Not applicable | Configuration of static validation test data node. See [Static Validation](#static-validation)
```

- feedback_dir:
  a string which specifies the name of a "feedback directory" in which the validator can produce "feedback files" in order to report additional information on the validation of the output file.

>   The feedbackdir must end with a path separator (typically '/' or '\\' depending on operating system),

Reviewer 2: ```suggestion
  feedback_dir must end with a path separator (typically '/' or '\\' depending on operating system),
```

They are chosen to minimize the configuration needed for a typical IOI-style problem, where points are given for passing all test cases in test data groups called subtasks.
To achieve that behavior, create groups `secret/subtask1`, `secret/subtask2`, ... and set the `score` value in the `secret/subtaskX/testdata.yaml`.

## Static Validator

### Overview

A static validator is a program that is given the submission files as input and can analyze the contents to accept or reject the submission.
Optionally, the static validator may assign a score to the submission.
By default there is no static validator.
A static validator may be provided under the `static_validator` directory, similar to a custom output validator.

### Static Validation Test Cases

Each test group may define a static validation test case.

> A static validation test case is defined within a group's `testdata.yaml` file by specifying the map `static_validation`.

Reviewer 2: Doesn't this mean that there's no way to have a static validator with no flags for pass-fail problems?
(Or alternatively: doesn't this force all pass-faill static validators to have `flags` set? Which feels unnecessary.)

Author: You can specify flags as the empty string. But perhaps having an enabled: true wouldn't be bad to be clear.

Reviewer 2: Alternatively, `static_validation: true` could be used when no relevant options should be set. (I'm not sure if that is more or less ugly than having an `enabled` flag...)

Author: I have no strong opinion, we use shorthands in YAML in many places so just true is fine here I think.

Author: Setting false would explicitly disable it then, which is essentially a no-op

The map can have two keys `flags`, and in the case of scoring test groups, `score`.
The key `flags` maps to a string which represents the additional arguments passed to the static validator in this group's static validation test case.
The key `score` maps to a float which represents both the maximum score achievable for the static validation test case and the default score assigned in case the static validator accepts the submission for that test case.
The static validator can override this score by outputting a value to `score.txt` in the feedback directory, the same as an output validator.
Aggregation is then applied to the test case in the same manner as other test cases.
It is an error to assign a score in a pass-fail test group.

### Invocation

> When invoked the static validator will be passed at least two command line parameter.

Reviewer 3: ```suggestion
When invoked the static validator will be passed at least two command line parameters.
```

> When invoked the static validator will be passed at least two command line parameter.

Reviewer 2: ```suggestion
When invoked, the static validator will be passed at least two command line parameters.
```

The validator should be possible to use as follows on the command line:
```sh
<static_validator_program> submission_dir feedback_dir [additional_arguments]
```

The meaning of the parameters listed above are:

- submission_dir:

>   a string specifying the absolute path to a directory containing a copy of the submission

Reviewer 3: for single-file submissions, is this the path to the file directly? Otherwise, should we pass the entrypoint / main file?

Reviewer 2: > for single-file submissions, is this the path to the file directly?

As I understand it, no. I also think linking directly to the file will make things more complicated for very little gain.

> Otherwise, should we pass the entrypoint / main file?

That is probably a good idea in either case?

Author: Maybe we both pass in path to main file (code) and entrypoint (binary)?

Reviewer 2: I don't think the binary is needed?

I think the most obvious thing to do would be to pass in `submission_dir` and `entrypoint` (which would contain the name of the mainfile/mainclass (the exact meaning is language-dependent)).

If we think the binary is needed, I think the most stable (across languages) solution is to handle that similarly to the source code: passing in `compilation_dir` and `program` (I don't think the binary is needed though?)

Reviewer 2: Never mind — as per @evouga's comment below, `submission_dir` isn't needed.

I think `entrypoint` alone would suffice, but if not, `entrypoint` and `binary` (or something like that) would also make sense.

Author: Entrypoint in C/C++/Rust is the binary
Entrypoint in Java/Scala is the main class
Entrypoint in Python/Bash/PHP is the main code file

Are we sure entrypoint suffices?

Reviewer: > Are we sure entrypoint suffices?

I think it needs the language and the entrypoint.

Author: Language and entry point added instead

>   a string specifying the absolute path to a directory containing a copy of the submission

Reviewer 2: "Submission files" might be more clear.
```suggestion
  a string specifying the absolute path to a directory containing a copy of the submission files
```

- feedback_dir:
  a string which specifies the name of a "feedback directory" in which the validator can produce "feedback files" in order to report additional information on the validation of the submission.

>   The feedbackdir must end with a path separator (typically '/' or '\\' depending on operating system),

Reviewer 2: ```suggestion
  feedback_dir must end with a path separator (typically '/' or '\\' depending on operating system),
```
